content/booklets/adacore-technologies-for-airborne-software/analysis.rst

```diff
@@ -0,0 +1,3180 @@
+index `I` into the array, from the first position through the next-to-last, the value of the element at position `I+1` is at least as
+large as the element at position `I`. In the degenerate case where the array is either empty or contains a single element (i.e., when
+the range of `I` is empty) the `for all` condition is considered to be True.
+
+It's clear that the postcondition expression says nothing about how the procedure is implemented. It's not pseudo-code for an
+algorithm but rather a property of the procedure that will need to be verified. It's the formalization of a requirement that happens
+to use Ada syntax. Moreover, a postcondition can refer to the values of formal parameters and/or global data, both at the point of call and
+the point of return. (In the above example, the postcondition could be strengthened by specifying that the value of `My_Array` on return
+is a permutation, possibly the identity mapping, of the value on entry.)
+And a function postcondition can refer to the value being returned by the function.
+
+A subprogram can also have a precondition (a Boolean expression), which is a requirement that the caller needs to satisfy and that is
+assumed to be True by the called subprogram. For example, a function that returns the maximum value in an array of integers
+should have a precondition that the array is non-empty. The postcondition that is shown reflects the two properties that need to be met:
+
+* The function result is at least as large as each element in the array, and
+* The function result is present in the array
+
+.. code-block:: ada
+
+   type Some_Array is array (Positive range <>) of Integer;
+
+   function Maximum (My_Array : in Some_Array) return Integer
+    with Pre  => My_Array'Length > 0,
+         Post =>
+             (for all I in My_Array'Range =>
+                  Maximum'Result >= My_Array (I)) and
+             (for some I in My_Array'Range =>
+                  Maximum'Result = My_Array (I));
+
+.. index:: single: Ada language; Contract-based programming
+
+Preconditions and postconditions, and related features such as type invariants, are referred to collectively as
+"contract-based programming" and were introduced in the Ada 2012 version of the language. Based on the assertion policy (as specified by a pragma), the contracts can be checked at run-time, raising an exception on failure. They also support (but do not require) formal analysis, since the Ada syntax is the same as is used in SPARK. In SPARK the contracts are subject to additional restrictions (for example they must conform to the SPARK language subset). The contracts are then considered to be low-level requirements and verification cases at the same time, used by the SPARK proof technology for formal verification, for example to demonstrate that if a subprogram satisfies its precondition then on return it will satisfy its postcondition.
```

Comment:
My understanding is that "run-time" should be rather written "run time" in this specific case.